magma/circuit.py

```diff
@@ -112,22 +115,20 @@ def _finalize_displays(self):
             self.add_inline_verilog(*display.get_inline_verilog())
 
     def finalize(self, defn):
-        final_placer = self.placer.finalize(defn)
-        final_context = DefinitionContext(final_placer)
+        self.placer = self.placer.finalize(defn)
         if self._insert_default_log_level:
             self.add_inline_verilog(_DEFAULT_VERILOG_LOG_STR, {}, {})
-        # So displays can refer to open files
-        self._finalize_file_opens()
+        self._finalize_file_opens()  # so displays can refer to open files
         self._finalize_displays()
-        # Close after displays
-        self._finalize_file_close()
-
-        # inline logic may introduce instances
-        with _DefinitionContextManager(final_context):
+        self._finalize_file_close()  # close after displays
+        # Inline logic may introduce instances.
+        with _DefinitionContextManager(self):
             for format_str, format_args, symbol_table in self._inline_verilog:
                 process_inline_verilog(defn, format_str, format_args,
                                        symbol_table)
-        return final_context
+        for builder in self._builders:
+            inst = builder.finalize()
+            self.placer.place(inst)
```

Comment:
I wonder if this should go before the inline/display logic since they might introduce these constructs during the finalize?

Reply:
Done.